### ISSUE TITLE
not routing after successful login

### ISSUE THREAD

**Issue by suriyaJaay** [SinghDigamber/angularfirebase-authentication#3]
Hi,

am not able to route to dashboard even though login is successful.  once click on sign-in button. app is loading for seconds and remains in login page. but is supposed to go to dashboard.

could you pls how to resolve this.?

and one more thing i want to know, what is the purpose of firebase/app in service?  i dont find ' auth' in node_modules under firebase folder.

thank you

**Comment by billbarni** [comment]
Same issue here. Nothing happens on a successful login.

**Comment by Vixxd** [comment]
This is also happening to me, only with username + password login, not via a social provider. When using something like Google it redirects as expected.

**Comment by ofirSparka19** [comment]
There's a conflict with the VerifiedEmail field.
The "isSignedIn" condition is based on also on the Email Verification value.
If you'd check your firestore, you'd see that it's set to False from some reason.

So, You aren't being redirected to the Dashboard since you're basically not logged in.

**Comment by MihajloNesic** [comment]
You can do it this way.

```typescript
// Sign in with email/password
SignIn(email, password) {
    return this.afAuth.auth.signInWithEmailAndPassword(email, password)
    .then(() => this.SetUserData(result.user))
    .then((result) => this.router.navigate(['/dashboard']))
    .catch((error) => window.alert(error.message));
}
```

**Comment by blouflashdb** [comment]
> You can do it this way.
> 
> ```ts
> // Sign in with email/password
> SignIn(email, password) {
>     return this.afAuth.auth.signInWithEmailAndPassword(email, password)
>     .then((result) => this.SetUserData(result.user))
>     .then(() => this.router.navigate(['/dashboard']))
>     .catch((error) => window.alert(error.message));
> }
> ```

Still doesn't work for me :( 
Node Version: v12.14.1
Angular CLI Version: 9.0.2
Fire: 5.4.2

**Comment by MihajloNesic** [comment]
> Still doesn't work for me :(
> Node Version: v12.14.1
> Angular CLI Version: 9.0.2
> Fire: 5.4.2

Did you notice the slash before *dashboard*?

I am using lower versions tho.
Node: 10.16.3
Angular CLI: 8.3.25
Fire: 5.1.0

**Comment by blouflashdb** [comment]
> > Still doesn't work for me :(
> > Node Version: v12.14.1
> > Angular CLI Version: 9.0.2
> > Fire: 5.4.2
> 
> Did you notice the slash before _dashboard_?
> 
> I am using lower versions tho.
> Node: 10.16.3
> Angular CLI: 8.3.25
> Fire: 5.1.0
> 
> Edit: Tested with Fire 5.4.2, it works perfectly

Yes, I replaced my SignIn methode with yours. I guess its a bug in the Angular CLI. ngOnInit in the Dashboard is not being called for some reason.

**Comment by danmincu** [comment]
the reason is not routing is because of the fact that the email is not verified. should you log out and back in the user info is being refreshed. 
I fixed this problem in my forked branch and also added phone SMS login, custom claims, better flow etc check it out https://github.com/danmincu/angularfirebase-authentication

**Comment by JDmko** [comment]
Hello, I solved it with setInterval () in signIn(email, password)

` SignIn(email, password) {
    return this.afAuth.auth.signInWithEmailAndPassword(email, password)
      .then((result) => {
        this.ngZone.run(() => {
          let i = 0;
        const loginTime = setInterval (() => {
              if( i = 2 ){
                this.router.navigate(['/dashboard']);
                clearInterval(loginTime)
              } //console.log(i)
        i++
},1000) 
        });
        this.SetUserData(result.user);
      }).catch((error) => {
        window.alert(error.message)
      })
  }`

and secureInnerPage.guard.ts
`canActivate(
    next: ActivatedRouteSnapshot,
    state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
      if(this.authService.isLoggedIn){
        //window.alert("your text")
        this.router.navigate(['/dashboard'])
      }
    return true;
  }`

**Comment by solanky** [comment]
> Hello, I solved it with setInterval () in signIn(email, password)
> 
> ```js
> SignIn(email, password) {
>     return this.afAuth.auth.signInWithEmailAndPassword(email, password)
>       .then((result) => {
>         this.ngZone.run(() => {
>           let i = 0;
>         const loginTime = setInterval (() => {
>               if( i = 2 ){
>                 this.router.navigate(['/dashboard']);
>                 clearInterval(loginTime)
>               } //console.log(i)
>         i++
> },1000) 
>         });
>         this.SetUserData(result.user);
>       }).catch((error) => {
>         window.alert(error.message)
>       })
>   }
> ```
> 
> and secureInnerPage.guard.ts
> 
> ```js
> canActivate(
>     next: ActivatedRouteSnapshot,
>     state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
>       if(this.authService.isLoggedIn){
>         //window.alert("your text")
>         this.router.navigate(['/dashboard'])
>       }
>     return true;
>   }
> ```
In my case, Login was not happening on First click on login button, while the user token were being set successfully in localstorage. I used your method of setinterval and now it is working fine on first click. What is the reason for such behavior?

**Comment by foch01** [comment]
I have the same problem. I think that the SetInterval solution is not a viable solution over time.

**Comment by MihajloNesic** [comment]
This is how I did it
```typescript
signIn(email, password) {
    return this.afAuth.auth.signInWithEmailAndPassword(email, password)
    .then((result) => this.setUserData(result.user))
    .then(() => this.router.navigate(['/dashboard']));
}
```
My account has a verified email

**Comment by foch01** [comment]
Doesn't work for me :/

**Comment by danmincu** [comment]
@foch01 it's easy to debug why it isn't working for you; 
Routing after logging to dashboard is guarded by the [Authguard](https://github.com/SinghDigamber/angularfirebase-authentication/blob/a746fc1c0e6f993d7ea47d1a5ece460bbc122878/src/app/shared/routing/app-routing.module.ts#L20) that depends on [isLoggedIn()](https://github.com/SinghDigamber/angularfirebase-authentication/blob/a746fc1c0e6f993d7ea47d1a5ece460bbc122878/src/app/shared/services/auth.service.ts#L82); I bet that "! user.emailIsVerified==false" is the culprit.

**Comment by foch01** [comment]
Even removing the user.emailVerified condition! == false
it does not work.

My email address is verified and equal to true

**Comment by RodriFerretty** [comment]
You signIn and try to navigate to a guard protected route, thus the **canActivate** is used, but the **canActivate** uses **localStorage user** to determine if you can perform navigation and the **localStorage user** is set in the **`authState.subscribe`**, which is called **AFTER** **canActivate**.
A simple solution to prove this is just set local storage with user data inside signIn method before `this.router.navigate`. Then you'll see that canActivate is able to get the user from localStorage and can navigate. 
Idk if this is the best solution, I'm just learning Angular.

**Comment by IvanMtze** [comment]
What I did, was to change my canActivate function, so now it look like this:
`next: ActivatedRouteSnapshot,
    state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
    if(this.authService.isLoggedIn !== true) {
      this.router.navigate(['sign-in'])
    }
    return true;
  }`

then, i changed my isLogged function, so now it looks like this
`get isLoggedIn(): boolean {
    const user = JSON.parse(localStorage.getItem('user'));
    return (user === null) ? false : true;
  }`

Remenber that you have to check that your function is saving locally the user in json format.
Actually, i just removed the verificateEmail function, it didn't work for me.

**Comment by stayacid** [comment]
Worked solution:
```ts
  SignIn(email: string, password: string) {
    return this.afAuth.signInWithEmailAndPassword(email, password)
      .then((result) => {
        this.SetUserData(result.user);
        this.afAuth.authState.subscribe((user) => {
          if (user) {
            this.router.navigate(['dashboard']);
          }
        })
      })
      .catch((error) => {
        window.alert(error.message)
      })
  }
```

**Comment by omararturo** [comment]
Solution:

```
// Auth logic to run auth providers
  async authLogin(provider: auth.AuthProvider) {
    try {
      await this.afAuth.signInWithPopup(provider).then(value => {
        this.setUserData(value.user).then(() => {
          this.ngZone.run(() => {
            this.router.navigate(['dashboard']);
          });
        });
      });
    }
    catch (error) {
      window.alert(error);
    }
  }
```

**Comment by lepatlamass** [comment]
> Solution:
> 
> ```ts
>   SignIn(email: string, password: string) {
>     return this.afAuth.signInWithEmailAndPassword(email, password)
>       .then((result) => {
>         this.SetUserData(result.user);
>         this.afAuth.authState.subscribe((user) => {
>           if (user) {
>             this.router.navigate(['dashboard']);
>           }
>         })
>       })
>       .catch((error) => {
>         window.alert(error.message)
>       })
>   }
> ```

This your solution is the best just tried and it works perfectly

**Comment by cflorioluis** [comment]
> Solution:
> 
> ```ts
>   SignIn(email: string, password: string) {
>     return this.afAuth.signInWithEmailAndPassword(email, password)
>       .then((result) => {
>         this.SetUserData(result.user);
>         this.afAuth.authState.subscribe((user) => {
>           if (user) {
>             this.router.navigate(['dashboard']);
>           }
>         })
>       })
>       .catch((error) => {
>         window.alert(error.message)
>       })
>   }
> ```

perfect, thanks
